compiler/qsc_qasm3/src/parser/stmt.rs

```diff
@@ -1048,9 +1070,13 @@ fn parse_end_stmt(s: &mut ParserContext) -> Result<EndStmt> {
 }
 
 /// Grammar: `expression SEMICOLON`.
-fn parse_expression_stmt(s: &mut ParserContext) -> Result<ExprStmt> {
+fn parse_expression_stmt(s: &mut ParserContext, lhs: Option<Expr>) -> Result<ExprStmt> {
     let lo = s.peek().span.lo;
-    let expr = expr::expr(s)?;
+    let expr = if let Some(lhs) = lhs {
+        expr::expr_with_lhs(s, lhs)?
+    } else {
+        expr::expr(s)?
+    };
     recovering_semi(s);
     Ok(ExprStmt {
         span: s.span(lo),
```

Comment:
I think the spans should come from the resulting expression:

```suggestion
    let expr = if let Some(lhs) = lhs {
        expr::expr_with_lhs(s, lhs)?
    } else {
        expr::expr(s)?
    };
    recovering_semi(s);
    Ok(ExprStmt {
        span: s.span(expr.span.lo),
```